lib/membrane_h264_plugin/parser/nalu_parser.ex

```diff
@@ -27,86 +25,68 @@ defmodule Membrane.H264.Parser.NALuParser do
+  @spec parse(binary(), t()) :: {NALu.t(), t()}
   def parse(
-        buffer,
-        state,
-        should_skip_last_nalu? \\ true
+        nalu_payload,
+        state
       ) do
-    payload = buffer.payload
-
-    {nalus, {scheme_parser_state, has_seen_keyframe?}} =
-      payload
-      |> NALuSplitter.extract_nalus(
-        buffer.pts,
-        buffer.dts,
-        state.prev_pts,
-        state.prev_dts,
-        should_skip_last_nalu?
-      )
-      |> Enum.map_reduce(
-        {state.scheme_parser_state, state.has_seen_keyframe?},
-        fn nalu, {scheme_parser_state, has_seen_keyframe?} ->
-          prefix_length = nalu.prefix_length
-
-          <<_prefix::binary-size(prefix_length), nalu_header::binary-size(1), nalu_body::binary>> =
-            nalu.payload
-
-          new_scheme_parser_state = SchemeParser.new(scheme_parser_state)
-
-          {parsed_fields, scheme_parser_state} =
-            SchemeParser.parse_with_scheme(
-              nalu_header,
-              Schemes.NALuHeader.scheme(),
-              new_scheme_parser_state
-            )
-
-          type = NALuTypes.get_type(parsed_fields.nal_unit_type)
-
-          try do
-            {parsed_fields, scheme_parser_state} =
-              parse_proper_nalu_type(nalu_body, scheme_parser_state, type)
-
-            status = if type != :non_idr or has_seen_keyframe?, do: :valid, else: :error
-            has_seen_keyframe? = has_seen_keyframe? or type == :idr
-
-            {%NALu{nalu | parsed_fields: parsed_fields, type: type, status: status},
-             {scheme_parser_state, has_seen_keyframe?}}
-          catch
-            "Cannot load information from SPS" ->
-              {%NALu{nalu | parsed_fields: parsed_fields, type: type, status: :error},
-               {scheme_parser_state, has_seen_keyframe?}}
-          end
-        end
+    prefix_length = (nalu_payload |> :binary.bin_to_list() |> Enum.find_index(&(&1 == 1))) + 1
+
+    <<_prefix::binary-size(prefix_length), nalu_header::binary-size(1), nalu_body::binary>> =
+      nalu_payload
+
+    new_scheme_parser_state = SchemeParser.new(state.scheme_parser_state)
+
+    {parsed_fields, scheme_parser_state} =
+      SchemeParser.parse_with_scheme(
+        nalu_header,
+        Schemes.NALuHeader.scheme(),
+        new_scheme_parser_state
       )
 
+    type = NALuTypes.get_type(parsed_fields.nal_unit_type)
+
+    {nalu, {scheme_parser_state, has_seen_keyframe?}} =
+      try do
+        {parsed_fields, scheme_parser_state} =
+          parse_proper_nalu_type(nalu_body, scheme_parser_state, type)
+
+        status = if type != :non_idr or state.has_seen_keyframe?, do: :valid, else: :error
+        has_seen_keyframe? = state.has_seen_keyframe? or type == :idr
+
+        {%NALu{
+           parsed_fields: parsed_fields,
+           type: type,
+           status: status,
+           prefix_length: prefix_length,
+           payload: nalu_payload
+         }, {scheme_parser_state, has_seen_keyframe?}}
```

Comment:
We can flatten this `{nalu, scheme_parser_state, has_seen_keyframe?}`

Reply:
done